connectivity/check/features.go

```diff
@@ -0,0 +1,217 @@
+// turned on or off (c.f. Enabled), while others additionally might include a
+// Mode string which provides more information about in what mode a
+// particular feature is running ((e.g. when running with CNI chaining,
+// Enabled will be true, and the Mode string will additionally contain the name
+// of the chained CNI).
+type FeatureStatus struct {
+	Enabled bool
+	Mode    string
+}
+
+func (s FeatureStatus) String() string {
+	str := "Disabled"
+	if s.Enabled {
+		str = "Enabled"
+	}
+
+	if len(s.Mode) == 0 {
+		return str
+	}
+
+	return fmt.Sprintf("%s:%s", str, s.Mode)
+}
+
+// FeatureSet contains the status
+type FeatureSet map[Feature]FeatureStatus
+
+// MatchRequirements returns true if the FeatureSet fs satisfies all the
+// requirements in reqs
+func (fs FeatureSet) MatchRequirements(reqs ...FeatureRequirement) bool {
+	for _, req := range reqs {
+		status := fs[req.feature]
+		if req.requiresEnabled && (req.enabled != status.Enabled) {
+			return false
+		}
+		if req.requiresMode && (req.mode != status.Mode) {
+			return false
+		}
+	}
+
+	// empty requirement list matches any feature set
+	return len(reqs) == 0
+}
+
+// FeatureRequirement defines a test requirement. A given FeatureSet may or
+// may not satisfy this requirement
+type FeatureRequirement struct {
+	feature Feature
+
+	requiresEnabled bool
+	enabled         bool
+
+	requiresMode bool
+	mode         string
+}
+
+// RequireFeatureEnabled constructs a FeatureRequirement which expects the
+// feature to be enabled
+func RequireFeatureEnabled(feature Feature) FeatureRequirement {
+	return FeatureRequirement{
+		feature:         feature,
+		requiresEnabled: true,
+		enabled:         true,
+	}
+}
+
+// RequireFeatureMode constructs a FeatureRequirement which expects the feature
+// to be in the given mode
+func RequireFeatureMode(feature Feature, mode string) FeatureRequirement {
+	return FeatureRequirement{
+		feature:      feature,
+		requiresMode: true,
+		mode:         mode,
+	}
+}
+
+func parseBoolStatus(s string) bool {
+	switch s {
+	case "Enabled", "enabled", "True", "true":
+		return true
+	}
+
+	return false
+}
+
+func (ct *ConnectivityTest) extractFeaturesFromConfigMap(ctx context.Context, client *k8s.Client, result FeatureSet) error {
+	cm, err := client.GetConfigMap(ctx, ct.params.CiliumNamespace, defaults.ConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to retrieve ConfigMap %q: %w", defaults.ConfigMapName, err)
+	}
+	if cm.Data == nil {
+		return fmt.Errorf("ConfigMap %q does not contain any configuration", defaults.ConfigMapName)
+	}
+
+	// Monitor aggregation level defaults to none.
+	mode := defaults.ConfigMapValueMonitorAggregatonNone
+	if v, ok := cm.Data[defaults.ConfigMapKeyMonitorAggregation]; ok {
+		mode = strings.ToLower(v)
+	}
+	result[FeatureMonitorAggregation] = FeatureStatus{
+		Enabled: mode != defaults.ConfigMapValueMonitorAggregatonNone,
+		Mode:    mode,
+	}
```

Comment:
snit: given monitor aggregation mode detection is now nicely contained within this function, I wonder whether it's still worth to keep `defaults.ConfigMapKeyMonitorAggregation` and `defaults.ConfigMapValueMonitorAggregatonNone` or rather just inline these consts

```suggestion
	mode := "none"
	if v, ok := cm.Data["monitor-aggregation"]; ok {
		mode = strings.ToLower(v)
	}
	result[FeatureMonitorAggregation] = FeatureStatus{
		Enabled: mode != "none",
		Mode:    mode,
	}
```

IMO this makes it slighty easier to read.